### PR TITLE
Fields dump load

### PR DESCRIPTION
**Pull request by kkg4theweb**
First cut of a PR to add support for dumping and (re)loading the 'fields' state.

### CODE REVIEW

File: python/tests/test_simulation.py [hunk 1]
```diff
@@ -336,6 +340,9 @@ def get_ref_field_point(sim):
         if chunk_sim:
             chunk_layout = sim1
 
+        fields_dump_fn = os.path.join(self.temp_dir, 'test_load_dump_fields.h5')
```
Comment (ahoenselaar) [inline python/tests/test_simulation.py:343]: Revert all diff in this file except for this section.
There should probably an assertion as well.

File: src/fields_dump.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+#include <string.h>
+
+#include <cassert>
+
+#include "meep.hpp"
+#include "meep_internals.hpp"
+
+namespace meep {
+
+#define DUMP_FIELD(field_name)                                              \
```
Comment (ahoenselaar) [inline src/fields_dump.cpp:33]: Can we avoid macros in all new code?

File: src/fields_dump.cpp [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  {                                                                         \
+    /*                                                                      \
+     * make/save a num_chunks x NUM_FIELD_COMPONENTS x 2 array counting     \
+     * the number of entries in the 'field_name' array for each chunk.      \
+     */                                                                     \
+    size_t *num_f_ = new size_t[num_chunks * NUM_FIELD_COMPONENTS * 2];     \
```
Comment (ahoenselaar) [inline src/fields_dump.cpp:39]: Can we please avoid C-style manual allocation in new code?

File: python/tests/test_simulation.py (hunk 1 above)
Comment (kkg4theweb) [reply python/tests/test_simulation.py:339]: Done.

File: src/fields_dump.cpp (hunk 2 above)
Comment (kkg4theweb) [reply src/fields_dump.cpp]: done

File: src/fields_dump.cpp (hunk 3 above)
Comment (kkg4theweb) [reply src/fields_dump.cpp]: done.

**Comment by stevengj** [comment]
It's fine to write everything to separate files (one per process, say).  You'll want to do a couple of things:

* Modify `save_dft_hdf5` to take an additional flag `single_parallel_file`.  If that's `false`, simply skip the parallel reductions in `dft_chunks_Ntotal`.   
* In your own code, follow the same two-pass structure: first compute the total size of the data on your process, then create the dataset, then write the chunks into that dataset.    Again, if ``single_parallel_file==true` we can do an optional parallel reduction after the first pass, and otherwise the code is the same.

Basically, to use HDF5 parallel I/O with a single file you have to create the file and create the dataset collectively, and then each process can call `write_chunk` as many times as it wants to write its own data into the file (as long as it writes into non-overlapping portions of the file).  That's why the `save_dft_hdf5` and `structure::dump` functions are organized the way they are.

**Comment by stevengj** [comment]
Possible Python API: to re-start a simulation, run the *same* simulation script, creating the `Simulation` object as usual with all of the geometry and sources etcetera, but before you do `sim.run`, first run a `sim.load(config)` function (that loads the fields, DFT, etcetera).

`config` would be some kind of argument that says what to load.  Maybe a JSON file, or just a `dict`, whose files give the various filenames?

**Comment by stevengj** [comment]
Closed by #1738?
